[NVPTX] Improve NVVMReflect Efficiency

YonahGoldberg opened the pull request:
The NVVMReflect pass simply replaces calls to nvvm-reflect functions with the appropriate constant, either the architecture number, or nvvm-reflect-ftz, found in the module's metadata.

The implementation is inefficient and does this by traversing through all instructions to find calls. The common case is that you never call nvvm-reflect, so this traversal is costly.

This PR:
- Updates the pass so that it finds the reflect functions by name, and then traverses through their uses to find the calls directly. 
- Adds a line (245) to make sure the dead nvvm-reflect definitions are erased. 
- Adds the ability to set reflect values via command line

YonahGoldberg committed:
making nvvm reflect more efficient

YonahGoldberg commented:
@AlexMaclean @Artem-B

YonahGoldberg committed:
cleanup

justinfargnoli commented on llvm/lib/Target/NVPTX/NVVMReflect.cpp:
```diff
@@ -208,23 +235,45 @@ static bool runNVVMReflect(Function &F, unsigned SmVersion) {
+PreservedAnalyses NVVMReflectPass::run(Module &M, ModuleAnalysisManager &AM) {
+  return NVVMReflect(VarMap).runOnModule(M) ? PreservedAnalyses::none()
+                                            : PreservedAnalyses::all();
+}
```
newline :)

YonahGoldberg replied:
fixed

YonahGoldberg committed:
newline

AlexMaclean commented on llvm/lib/Target/NVPTX/NVVMReflect.cpp:
```diff
@@ -208,23 +235,45 @@ static bool runNVVMReflect(Function &F, unsigned SmVersion) {
-NVVMReflectPass::NVVMReflectPass() : NVVMReflectPass(0) {}
+  bool Changed = false;
+  // Names of reflect function to find and replace
+  SmallVector<std::string, 3> ReflectNames = {
```
I think you can use an array of StringRef here to avoid allocating these strings.

YonahGoldberg replied:
Yup, fixed that.

AlexMaclean commented on llvm/lib/Target/NVPTX/NVVMReflect.cpp:
```diff
@@ -78,27 +91,51 @@ INITIALIZE_PASS(NVVMReflect, "nvvm-reflect",
+  assert(F->getReturnType()->isIntegerTy() &&
+         "_reflect's return type should be integer");
 
   SmallVector<Instruction *, 4> ToRemove;
   SmallVector<Instruction *, 4> ToSimplify;
```
My guess is that doing removal and simplification at the end was to avoid disrupting the iteration over instructions in the function. Now that we're iterating over uses I don't think these data structurers are needed.

AlexMaclean commented on llvm/lib/Target/NVPTX/NVVMReflect.cpp:
```diff
@@ -39,33 +38,47 @@
 #include "llvm/Transforms/Scalar.h"
 #include "llvm/Transforms/Utils/BasicBlockUtils.h"
 #include "llvm/Transforms/Utils/Local.h"
+#include "llvm/Transforms/Utils/StripGCRelocates.h"
```
Is this being used somewhere?

YonahGoldberg replied:
No, idk how that got added. Deleted it.

AlexMaclean reviewed (commented):
Looks like some nice cleanup. Please add tests for the new debugging option.

Artem-B commented:
@AlexMaclean do you think we could reuse intrinsic autoupgrade machinery for this, instead of making reflect processing more complicated?

It could be somewhat useful for other purposes.
E.g. we could introduce a const-foldable (when we know the GPU we're targeting) `nvvm.arch()` which would return __CUDA_ARCH__ value and upgrade nvvm_reflect to it. Bonus point is that it would also be useful for IR users to parametrize their code without relying on NVVMReflect.

@jhuber6 would something like that help with some of your offloading cases. I recall you did run into trouble with NVVMReflect a while back.

jhuber6 commented:
> @AlexMaclean do you think we could reuse intrinsic autoupgrade machinery for this, instead of making reflect processing more complicated?
> 
> It could be somewhat useful for other purposes. E.g. we could introduce a const-foldable (when we know the GPU we're targeting) `nvvm.arch()` which would return **CUDA_ARCH** value and upgrade nvvm_reflect to it. Bonus point is that it would also be useful for IR users to parametrize their code without relying on NVVMReflect.
> 
> @jhuber6 would something like that help with some of your offloading cases. I recall you did run into trouble with NVVMReflect a while back.

All that really matters for correctness is that this pass is always run and it always does constant prop + DCE when it's used as an edge directly.

AlexMaclean commented:
> @AlexMaclean do you think we could reuse intrinsic autoupgrade machinery for this, instead of making reflect processing more complicated?
> 
> It could be somewhat useful for other purposes. E.g. we could introduce a const-foldable (when we know the GPU we're targeting) `nvvm.arch()` which would return **CUDA_ARCH** value and upgrade nvvm_reflect to it. Bonus point is that it would also be useful for IR users to parametrize their code without relying on NVVMReflect.
> 
> @jhuber6 would something like that help with some of your offloading cases. I recall you did run into trouble with NVVMReflect a while back.

It's an interesting idea, and I agree that intrinsics seem marginally cleaner than relying on string arguments, but I don't think there will be much benefit in the end. For correctness, we'll still need something like NVVMReflect to run to fold these calls away and do the constant prop + DCE. 

I think with some cleanup this new approach will not be more complicated. In some ways I think this allows for simplification because we shouldn't need to be as careful about deleting and simplifying now that we're not iterating of instructions in a function.

Artem-B commented on llvm/lib/Target/NVPTX/NVVMReflect.cpp:
```diff
@@ -78,27 +91,51 @@ INITIALIZE_PASS(NVVMReflect, "nvvm-reflect",
-    return false;
+static cl::list<std::string>
+    ReflectList("nvvm-reflect-list", cl::value_desc("name=<int>"), cl::Hidden,
+                cl::desc("A list of string=num assignments"),
```
Nit: ".. list of comma-separated key=<int> values"

YonahGoldberg replied:
Fixed.

Artem-B commented on llvm/lib/Target/NVPTX/NVVMReflect.cpp:
```diff
@@ -78,27 +91,51 @@ INITIALIZE_PASS(NVVMReflect, "nvvm-reflect",
-    assert(F.getReturnType()->isIntegerTy() &&
-           "_reflect's return type should be integer");
-    return false;
+/// The command line can look as follows :
+/// -nvvm-reflect-list a=1,b=2 -nvvm-reflect-list c=3,d=0 -R e=2
+/// The strings "a=1,b=2", "c=3,d=0", "e=2" are available in the
+/// ReflectList vector. First, each of ReflectList[i] is 'split'
+/// using "," as the delimiter. Then each of this part is split
+/// using "=" as the delimiter.
```
I'd rephrase it in terms of what the option does. 

E.g. "Allow users to specify additional key/value pairs to reflect. Accepts a comma-separated list of <string>=<int> pairs."
Elaborate on what happens when the user-specified value overrides the "standard" ones.
Skip the details of how the option is processed. If it's needed it should be documented in the comments next to the implementation..

YonahGoldberg replied:
Moved implementation details into the body and added a higher level blurb like you described Artem above the option definition.

Artem-B commented on llvm/lib/Target/NVPTX/NVVMReflect.cpp:
```diff
@@ -78,27 +91,51 @@ INITIALIZE_PASS(NVVMReflect, "nvvm-reflect",
+
+  for (unsigned I = 0, E = ReflectList.size(); I != E; ++I) {
+    LLVM_DEBUG(dbgs() << "Option : " << ReflectList[I] << "\n");
+    SmallVector<StringRef, 4> NameValList;
+    StringRef(ReflectList[I]).split(NameValList, ",");
```
Do we need to implement the comma-splitting? Wouldn't cl::list handle multiple values for us already if the user specifies the option more than once? If cl::list already handles multiple values, then we should probably rename the option into `--nvvm-reflect-add` or something...

Also, using a more modern C++ features for iteration may make work well here.

E.g.
```
for (StringRef O: ReflectList) {
   auto [Key, Value] = O.split("=");
   ...
}
```

YonahGoldberg replied:
I made the splitting more modern.

For the comma-splitting, I think you need it. If you do ... --nvvm-reflect-add a=1,b=2 --nvm-reflect-add c=3,d=4,
then the reflect list looks like {"a=1,b=2", "c=3,d=4"}. So for each option specified you still need to split on the comma.

Artem-B replied:
One option ->  one key=value pair. There's no need to support specifying multiple pairs via a single option.

YonahGoldberg replied:
Is it fine instead to just add the `cl::CommaSeparated` argument when creating the option (see current version of code)? Now, like you said before, `cl::list` handles if the user specifies the option more than once and it splits on the commas for us.

Artem-B replied:
There's indeed more than one way to do things. The question is -- which one makes more sense?

The way I see it the CLI should be as simple to implement, and as straightforward to use as we can make it. From that standpoint along the comma separated list of key=value pairs will stand out as a sore thumb. In some cases we have to do that, but this is not one of them. The "standard" way for the options is to specify *one* thing per option. E.g we're not passing comma-separate list of macros to define with `-D`. Speaking of which, `-D` is actually semantically very close to what you're trying to do, as NVVMReflect is, effectively, a preprocessor for the IR, and your new option defines one of the macro values for it. 

So, let's keep thing simple, and define one reflect value at a time.

YonahGoldberg replied:
Ok, I fixed it so it has the desired behavior. Thanks for all the great feedback!!

jhuber6 commented:
It's also worth noting that https://github.com/llvm/llvm-project/pull/134016 is AMD's attempt at doing the same functionality.

jhuber6 commented on llvm/lib/Target/NVPTX/NVVMReflect.cpp:
```diff
@@ -4,6 +4,12 @@
 //
+// NVIDIA_COPYRIGHT_BEGIN
+//
+// Copyright (c) 2023-2025, NVIDIA CORPORATION.  All rights reserved.
```
Can we have other copyrights in LLVM code?

YonahGoldberg replied:
Oops! I didn't mean to add that. I'm pretty sure you guys don't have NVIDIA copyrights anywhere in the codebase, so it shouldn't be there.

YonahGoldberg committed:
reflect improvements

YonahGoldberg commented on llvm/lib/Target/NVPTX/NVVMReflect.cpp:
```diff
@@ -39,33 +38,47 @@
+  // __CUDA_FTZ is assigned in `runOnModule` by checking nvvm-reflect-ftz module
+  // metadata.
+  explicit NVVMReflect(unsigned int Sm) : ModulePass(ID) {
+    VarMap["__CUDA_ARCH"] = Sm * 10;
```
Fixed this.

YonahGoldberg committed:
comment move

YonahGoldberg commented on llvm/lib/Target/NVPTX/NVVMReflect.cpp:
```diff
@@ -78,27 +91,51 @@ INITIALIZE_PASS(NVVMReflect, "nvvm-reflect",
+          M.getModuleFlag("nvvm-reflect-ftz")))
+    VarMap["__CUDA_FTZ"] = Flag->getSExtValue();
+
+  for (unsigned I = 0, E = ReflectList.size(); I != E; ++I) {
```
Yes.

YonahGoldberg committed:
auto

YonahGoldberg commented on llvm/lib/Target/NVPTX/NVVMReflect.cpp:
```diff
@@ -78,27 +91,51 @@ INITIALIZE_PASS(NVVMReflect, "nvvm-reflect",
+    StringRef(ReflectList[I]).split(NameValList, ",");
+    for (unsigned J = 0, EJ = NameValList.size(); J != EJ; ++J) {
+      SmallVector<StringRef, 2> NameValPair;
+      NameValList[J].split(NameValPair, "=");
```
Fixed that.

YonahGoldberg committed:
remove nvidia copyright

YonahGoldberg committed:
improve error messages

YonahGoldberg committed:
fix command line options

YonahGoldberg committed:
fix command line options

YonahGoldberg committed:
final reflect cleanup

YonahGoldberg committed:
final reflect cleanup

YonahGoldberg committed:
clang format

YonahGoldberg committed:
forgot to set Changed variable

YonahGoldberg committed:
typo in comment

AlexMaclean commented on llvm/lib/Target/NVPTX/NVVMReflect.cpp:
```diff
@@ -39,186 +39,207 @@
+private:
+  // Map from reflect function call arguments to the value to replace the call
+  // with. Should include __CUDA_FTZ and __CUDA_ARCH values.
+  StringMap<int> ReflectMap;
```
In practice I think we are just going to be implicitly casting these ints to unsigned when calling `ConstantInt::get`. I think it might be better to avoid the signed -> unsigned conversion and make the type unsigned in this map.

AlexMaclean commented on llvm/lib/Target/NVPTX/NVVMReflect.cpp:
```diff
@@ -39,186 +39,207 @@
 namespace {
-class NVVMReflect : public FunctionPass {
+class NVVMReflect {
+private:
```
This private is redundant.

AlexMaclean commented on llvm/lib/Target/NVPTX/NVVMReflect.cpp:
```diff
@@ -39,186 +39,207 @@
+  bool runOnModule(Module &M);
+};
+} // namespace
 
-  bool runOnFunction(Function &) override;
+namespace {
```
Nit: just use a single anonymous namespace.

AlexMaclean commented on llvm/lib/Target/NVPTX/NVVMReflect.cpp:
```diff
@@ -39,186 +39,207 @@
+    LLVM_DEBUG(dbgs() << "ReflectOption : " << Option << "\n");
+    auto [Name, Val] = Option.split('=');
+    if (Name.empty())
+      report_fatal_error("Empty name in nvvm-reflect-list option '" + Option +
```
Update these error messages to use the correct name of the option also maybe try to squeeze the message onto one line of code if you can.

AlexMaclean commented on llvm/lib/Target/NVPTX/NVVMReflect.cpp:
```diff
@@ -39,186 +39,207 @@
+    // c"__CUDA_ARCH\00" call i32 @__nvvm_reflect(ptr addrspacecast (ptr
+    // addrspace(1) @arch to ptr)) We need to extract the string argument from
+    // the call (i.e. "__CUDA_ARCH")
+    if (!isa<CallInst>(U))
+      report_fatal_error(
+          "__nvvm_reflect can only be used in a call instruction");
+    CallInst *Call = cast<CallInst>(U);
```
I think a little bit more idiomatic way to do this would be to use a `dyn_cast` and then the error check can be on whether the result is null

AlexMaclean commented on llvm/lib/Target/NVPTX/NVVMReflect.cpp:
```diff
@@ -39,186 +39,207 @@
+    StringRef ReflectArg =
+        cast<ConstantDataSequential>(ConstantStr)->getAsString();
+    // Remove the null terminator from the string
     ReflectArg = ReflectArg.substr(0, ReflectArg.size() - 1);
```
While we're here you may as well use `drop_back()` to remove the null char.

AlexMaclean commented on llvm/lib/Target/NVPTX/NVVMReflect.cpp:
```diff
@@ -39,186 +39,207 @@
+  // Replace an instruction with a constant and add all users of the instruction
+  // to the worklist
+  auto ReplaceInstructionWithConst = [&](Instruction *I, Constant *C) {
+    for (User *U : I->users()) {
```
nit: remove the braces from this for loop

AlexMaclean commented on llvm/lib/Target/NVPTX/NVVMReflect.cpp:
```diff
@@ -39,186 +39,207 @@
+  while (!Worklist.empty()) {
+    Instruction *I = Worklist.pop_back_val();
+    if (Constant *C =
+            ConstantFoldInstruction(I, Call->getModule()->getDataLayout())) {
```
nit: pull the calls to get the DL out of this loop

AlexMaclean reviewed (commented):
Looking lots better, couple minor stylistic issues to fix

YonahGoldberg committed:
final cleanup

YonahGoldberg committed:
formatting

YonahGoldberg committed:
formatting

YonahGoldberg committed:
int -> unsigned

YonahGoldberg committed:
const correct probably

YonahGoldberg committed:
format

YonahGoldberg committed:
forgot a getOperand call

YonahGoldberg committed:
format

AlexMaclean commented on llvm/lib/Target/NVPTX/NVVMReflect.cpp:
```diff
@@ -39,186 +39,207 @@
+  SmallVector<Instruction *, 8> Worklist;
+  // Replace an instruction with a constant and add all users of the instruction
+  // to the worklist
+  auto ReplaceInstructionWithConst = [&](Instruction *const I,
+                                         Constant *const C) {
+    for (auto *const U : I->users())
+      if (auto *const UI = dyn_cast<Instruction>(U))
+        Worklist.push_back(UI);
```
While I'm generally in favor of adding `const` qualifiers, I think cases like this go a little too far. A situation like this is simple enough that I think `const` is adding unnecessary clutter and harming readability without any clear benefit. In general, I haven't seen many cases where `const` is worth adding after a pointer and a rule of thumb I follow is not to add it in these cases.

YonahGoldberg replied:
My preference would be to add `const` wherever I can. I agree it's a bit verbose, but the whole point is that if I see something `const`, I know it's immutable and if I see something not `const`, it likely mutates. If I only annotate some variables `const`, then this logic breaks.

For how much `const` is used in LLVM currently, it seems to be on a file-by-file basis, just from a simple keyword search. But for the files that do have it, it's usually pervasive throughout the file. There's nothing in the style guide that mentions it.

@Artem-B, what do you think, do you mind my pervasive use of const?--If you agree with Alex I'll go and fix it.

Artem-B replied:
I'm inclined to agree with @AlexMaclean that in this particular case `auto *const` is a bit of an overkill that just clutters the code without buying us much. 

In general, I would rather prefer using `const &` instead of raw pointers wherever possible. I suspect that alone will eliminate a fair number of const use instances we may want to argue about. If you don't expect the value to be nullptr (like iterating over collections, or passing something as a function argument) `const &` is the way to go.

In general, the code should be painfully obvious. If the trivial code raises questions, there should be a good reason for that cognitive overhead, and, preferably, those reasons should be already explained in the comments.

AlexMaclean commented on llvm/lib/Target/NVPTX/NVVMReflect.cpp:
```diff
@@ -39,186 +39,207 @@
+  // Map from reflect function call arguments to the value to replace the call
+  // with. Should include __CUDA_FTZ and __CUDA_ARCH values.
+  StringMap<unsigned> ReflectMap;
+  bool handleReflectFunction(Module &M, const StringRef ReflectName);
+  void populateReflectMap(Module &M);
+  void foldReflectCall(CallInst *const Call, Constant *const NewValue);
```
Even if you were to add `const` on the arguments in the definition of these functions, I don't think it makes sense on the definition. These `const`s are just an implementation detail indicating that in the body you won't be reassigning these variables, they add no information in the declaration.

AlexMaclean reviewed (commented):
Looks essentially good to go. As a stylistic matter I think the `const` qualifier has been somewhat over-applied. In general I think it makes sense to add `const` only when it improves readability, and not where ever possible.

YonahGoldberg committed:
format

YonahGoldberg committed:
newline

YonahGoldberg commented:
@Artem-B Currently I'm failing a Clang test because of the fact that I switched NVVMReflect to be a module pass now instead of a function pass. The test is `clang/test/CodeGen/builtins-nvptx.c: 

```
__device__ bool reflect() {

// CHECK: call i32 @llvm.nvvm.reflect(ptr {{.*}})

  unsigned x = __nvvm_reflect("__CUDA_ARCH");
  return x >= 700;
}
```

I'm failing now because the NVVMReflect pass is getting run and removing the call. The difference that's causing it to get run now is in `NVPTXTargetMachine.cpp`:

```
PB.registerPipelineStartEPCallback(
      [this](ModulePassManager &PM, OptimizationLevel Level) {
        // We do not want to fold out calls to nvvm.reflect early if the user
        // has not provided a target architecture just yet.
        if (Subtarget.hasTargetName())
          PM.addPass(NVVMReflectPass(Subtarget.getSmVersion()));

        FunctionPassManager FPM;
        // Note: NVVMIntrRangePass was causing numerical discrepancies at one
        // point, if issues crop up, consider disabling.
        FPM.addPass(NVVMIntrRangePass());
        if (EarlyByValArgsCopy)
          FPM.addPass(NVPTXCopyByValArgsPass());
        PM.addPass(createModuleToFunctionPassAdaptor(std::move(FPM)));
      });
```

I changed this in my PR so that the reflect pass is added to the module pass manager, not the function pass manager. For some reason, these function passes added in the callback are NOT run in the lit-test, but the module pass is.

For reference, the clang cmd run is: `clang -cc1 -internal-isystem /mnt/data/ygoldberg/mainline/llvm-project/build/lib/clang/21/include -nostdsysteminc -ffp-contract=off -triple nvptx64-unknown-unknown -target-cpu sm_70 -target-feature +ptx63 -fcuda-is-device -emit-llvm -o - -x cuda /mnt/data/ygoldberg/mainline/llvm-project/clang/test/CodeGen/builtins-nvptx.c`

Do you know if this is an issue or why this is happening?

Artem-B commented:
I don't know why there's a difference in whether reflect gets to run or not while emitting unoptimized IR. 
Can you try adding `-disable-llvm-optzns` and see if that helps?

In the end it does not really matter much. We can adjust the test to verify that we see the intended reflect value in the generated IR.

YonahGoldberg commented:
This option worked. If I add it to all the clang compile commands for the `builtins-nvptx.c`, all tests pass. Should we just go ahead and commit that instead of changing the reflect test?

YonahGoldberg committed:
removing const

YonahGoldberg committed:
format

AlexMaclean reviewed (approved):
LGTM, please wait for @Artem-B's approval before landing.

Artem-B commented on llvm/lib/Target/NVPTX/NVVMReflect.cpp:
```diff
@@ -39,186 +39,201 @@
+PreservedAnalyses NVVMReflectPass::run(Module &M, ModuleAnalysisManager &AM) {
+  return NVVMReflect(SmVersion).runOnModule(M) ? PreservedAnalyses::none()
+                                               : PreservedAnalyses::all();
+}
```
Nit: missing newline.

Artem-B commented on llvm/test/CodeGen/NVPTX/nvvm-reflect-options.ll:
```diff
@@ -0,0 +1,26 @@
+; Verify that when passing in command-line options to NVVMReflect, that reflect calls are replaced with
```
The test is functionally fine, but it also makes me stop and think "what exactly are we doing here and why?".

Two points:
- generating test source for such a small test adds as much complexity as would be adding a copy of the file with additional module options. 
- we're checking for both reflected values encoded into one return value. I would keep them in separate functions, checked individually.
- we're only checking ability to override the values already known to the reflect pass. We need a test case that reflects a non-standard value, and verify what happens when it's passed via command line and when it's not.
- automating the checks would be beneficial as it's sometimes useful to see more of the generated IR than just a single instruction.

Artem-B commented on llvm/test/CodeGen/NVPTX/nvvm-reflect-options.ll:
```diff
@@ -0,0 +1,26 @@
+@ftz = private unnamed_addr addrspace(1) constant [11 x i8] c"__CUDA_FTZ\00"
+@arch = private unnamed_addr addrspace(1) constant [12 x i8] c"__CUDA_ARCH\00"
+
+; RUN: opt -passes=nvvm-reflect -mtriple=nvptx-nvidia-cuda -nvvm-reflect-add __CUDA_FTZ=1 -nvvm-reflect-add __CUDA_ARCH=350 %s -S | FileCheck %s --check-prefix=CHECK-FTZ1-ARCH350
+; RUN: opt -passes=nvvm-reflect -mtriple=nvptx-nvidia-cuda -nvvm-reflect-add __CUDA_FTZ=0 -nvvm-reflect-add __CUDA_ARCH=520 %s -S | FileCheck %s --check-prefix=CHECK-FTZ0-ARCH520
+
+; Verify that if we have module metadata that sets __CUDA_FTZ=1, that gets overridden by the command line arguments
+
+; RUN: cat %s > %t.options
+; RUN: echo '!llvm.module.flags = !{!0}' >> %t.options
+; RUN: echo '!0 = !{i32 4, !"nvvm-reflect-ftz", i32 1}' >> %t.options
+; RUN: opt -passes=nvvm-reflect -mtriple=nvptx-nvidia-cuda -nvvm-reflect-add __CUDA_FTZ=0 -nvvm-reflect-add __CUDA_ARCH=520 %t.options -S | FileCheck %s --check-prefix=CHECK-FTZ0-ARCH520
```
I'd move all RUN lines to the top of the file, and update the comments about construction of `%t.options` making it clear that it's the same test with module-scope options added. I.e. instead of just saying "Verify that options get overriden" rephrase it along the lines of "To test our ability to override module options, create a copy of this test file with module options appended."

Artem-B reviewed (approved):
LGTM for the code. Tests could use a bit more polishing.

YonahGoldberg committed:
updated reflect options test

YonahGoldberg committed:
newline

Artem-B reviewed (commented):
Almost there. Few more test nits.

YonahGoldberg committed:
final style changes

YonahGoldberg commented on llvm/test/CodeGen/NVPTX/nvvm-reflect-options.ll:
```diff
@@ -1,26 +1,53 @@
+; CHECK-FTZ1: define i32 @test_ftz()
+; CHECK-FTZ1: ret i32 1
+; CHECK-FTZ0: define i32 @test_ftz()
+; CHECK-FTZ0: ret i32 0
```
fixed.

YonahGoldberg commented on llvm/test/CodeGen/NVPTX/nvvm-reflect-options.ll:
```diff
@@ -1,26 +1,53 @@
+; that aren't __CUDA_ARCH or __CUDA_FTZ. If that argument is given a value on the command-line, the reflect call should be replaced with that value.
+; Otherwise, the reflect call should be replaced with 0.
+
+; RUN: opt -passes=nvvm-reflect -mtriple=nvptx-nvidia-cuda -nvvm-reflect-add __CUDA_FTZ=1 -nvvm-reflect-add __CUDA_ARCH=350 %s -S | FileCheck %s --check-prefix=CHECK-FTZ1 --check-prefix=CHECK-ARCH350 --check-prefix=CHECK-CUSTOM-ABSENT
```
fixed.

YonahGoldberg commented on llvm/test/CodeGen/NVPTX/nvvm-reflect-options.ll:
```diff
@@ -1,26 +1,53 @@
+; Otherwise, the reflect call should be replaced with 0.
+
+; RUN: opt -passes=nvvm-reflect -mtriple=nvptx-nvidia-cuda -nvvm-reflect-add __CUDA_FTZ=1 -nvvm-reflect-add __CUDA_ARCH=350 %s -S | FileCheck %s --check-prefix=CHECK-FTZ1 --check-prefix=CHECK-ARCH350 --check-prefix=CHECK-CUSTOM-ABSENT
+; RUN: opt -passes=nvvm-reflect -mtriple=nvptx-nvidia-cuda -nvvm-reflect-add __CUDA_FTZ=0 -nvvm-reflect-add __CUDA_ARCH=520 %s -S | FileCheck %s --check-prefix=CHECK-FTZ0 --check-prefix=CHECK-ARCH520 --check-prefix=CHECK-CUSTOM-ABSENT
```
fixed.

Artem-B reviewed (approved):
LGTM.

YonahGoldberg committed:
clang format

YonahGoldberg committed:
Merge branch 'main' into nvvm-reflect

AlexMaclean commented:
Merging on behalf of @YonahGoldberg at his request offline.
